### PR TITLE
🔨 Support for tools for data utils

### LINKED ISSUE #2429: Function Calling Fine-tuning Support
Dear trl team,

As function calling fine-tuning becomes increasingly important for teaching language models to use tools effectively in agentic workflows, I'm pleased to see Hugging Face has unified the encoding and processing of various tool-calling messages through chat templates (https://huggingface.co/blog/unified-tool-use) .

I would like to inquire about trl's current or planned support for function calling fine-tuning. I haven't found related documentation in SFTTrainer (the conversation data format fields only include user and assistant roles). I believe supporting function calling would require at least these features:

1. Support for processing and encoding messages with user, assistant, and tool roles (already supported by the tokenizer library)
2. Loss calculation specifically for assistant parts (including both general messages and tool-calling messages)

Alternatively, is it possible to implement function calling fine-tuning using existing SFTTrainer features? For example, through a custom collate_fn similar to vlm_sft?

Thank you for your clarification!

### PR DESCRIPTION
**Pull request by August-murr**
# What does this PR do?

Fixes #2429 
Draft a pull request to add support for function calling fine-tuning.
I suggest including a [tool-using dataset](https://huggingface.co/datasets/August4293/tool_sample_dataset) in the internal testing datasets for future development.

Here's example code to test the changes.

```python
dataset = load_dataset("August4293/tool_sample_dataset", split="train")
model_name = "trl-internal-testing/tiny-LlamaForCausalLM-3"
model = AutoModelForCausalLM.from_pretrained(model_name)
processor = AutoProcessor.from_pretrained("NousResearch/Hermes-2-Pro-Llama-3-8B")#loading llama 3 tokenizer
#tools
def get_current_temperature(location: str):
    """
    Gets the temperature at a given location.

    Args:
        location: The location to get the temperature for
    """
    return 22.0

tools = [get_current_temperature]

# Initialize the SFTConfig
config = SFTConfig(output_dir="/dir",
                   dataset_text_field="messages",
                   tools=[get_current_temperature]
                   )
# Initialize the SFTTrainer
trainer = SFTTrainer(
    model=model,
    train_dataset=dataset,
    processing_class=processor,
    args=config
)
prepared_dataset = trainer.train_dataset
print("Prepared dataset:", processor.decode(prepared_dataset["input_ids"][0])) #decode and print to see an example
```
which adds a system prompt to the dataset along with the schema of the functions.

## Trainer Implementations
- [x] `SFTTrainer`
- [ ] `DPOTrainer`

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/trl/tree/main/docs).
- [ ] Did you write any new necessary tests?

## Who can review?

@qgallouedec

### CODE REVIEW

**Comment by August-murr** [comment]
@qgallouedec how's everything so far?
Is there anything you'd like me to change?

**Comment by qgallouedec** [comment]
Looks good! We just need to update the docstrings of the functions and add some unittests

**Comment by August-murr** [comment]
I'm assuming we should also integrate the functions from `data_utils.py` into all the trainers, correct?

**Comment by qgallouedec** [comment]
> I'm assuming we should also integrate the functions from `data_utils.py` into all the trainers, correct?

Indeed, but we'll do that in follow-up PR. I think it's the best way to go

File: tests/test_data_utils.py [hunk 1]
```diff
@@ -195,6 +195,33 @@ def test_maybe_apply_chat_template(self, tokenizer_id, example):
             self.assertIsInstance(result["label"], bool)
             self.assertEqual(result["label"], example["label"])
 
+    def test_apply_chat_template_with_tools(self):
+        tokenizer = AutoProcessor.from_pretrained("NousResearch/Hermes-2-Pro-Llama-3-8B")
```
Comment (qgallouedec) [inline tests/test_data_utils.py:199]: Can we replace this by any of our [tiny models](https://huggingface.co/collections/trl-internal-testing/tiny-models-674b1f53bd1085e7d3c20fde)?

File: trl/trainer/sft_config.py [hunk 2]
```diff
@@ -30,6 +30,9 @@ class SFTConfig(TrainingArguments):
         dataset_text_field (`str`, *optional*, defaults to `"text"`):
             Name of the text field of the dataset. If provided, the trainer will automatically create a
             [`ConstantLengthDataset`] based on `dataset_text_field`.
+        tools (`Optional[list]`, *optional*, defaults to `None`):
+            List of functions that define the available tools for a function calling model.
+            The trainer uses these functions to generate a schema that gets incorporated into the system prompt.
```
Comment (qgallouedec) [inline trl/trainer/sft_config.py:35]:
```suggestion
```

Let's keep trainers unchanged for this PR

File: trl/data_utils.py [hunk 3]
```diff
@@ -60,9 +60,9 @@ def is_conversational(example: dict[str, Any]) -> bool:
     return False
 
 
-def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer) -> dict[str, str]:
+def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer, tools: Optional[list[dict]] = None) -> dict[str, str]:
     r"""
-    Apply a chat template to a conversational example.
+    Apply a chat template to a conversational example along with the schema for a list of functions in `tools`.
 
     For more details, see [`maybe_apply_chat_template`].
```
Comment (qgallouedec) [inline trl/data_utils.py:67]: You should also update `maybe_apply_chat_template`

File: trl/data_utils.py [hunk 4]
```diff
@@ -60,9 +60,9 @@ def is_conversational(example: dict[str, Any]) -> bool:
     return False
 
 
-def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer) -> dict[str, str]:
+def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer, tools: Optional[list[dict]] = None) -> dict[str, str]:
```
Comment (qgallouedec) [inline trl/data_utils.py:63]: is it really a list of dicts? It seems to be a list of callables instead.

Comment (August-murr) [reply trl/data_utils.py:63]: I thought that too but then [the docs](https://huggingface.co/docs/transformers/main/en/internal/tokenization_utils#transformers.PreTrainedTokenizerBase.apply_chat_template)

> tools (List[Dict], optional) — A list of tools (callable functions) that will be accessible to the model.

I think the reason is that it can be both a list of python functions or the json schema of the function as a dict.

Comment (qgallouedec) [reply trl/data_utils.py:63]: Interesting, @Rocketknight1 can you take a look?

Comment (Rocketknight1) [reply trl/data_utils.py:63]: Hey! The answer is that tools can be both. Tools can be either a `Callable` function or a JSON schema `Dict`. If a function is passed, we generate the schema dict for the user - the details are [here](https://huggingface.co/docs/transformers/main/en/chat_templating#understanding-tool-schemas).

The type hints for `apply_chat_template` need to be updated, though - it should be `List[Union[Dict, Callable]]`. I'll make a PR for that now!

Comment (qgallouedec) [reply trl/data_utils.py:63]:
```suggestion
def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer, tools: Optional[list[Union[dict, Callable]]] = None) -> dict[str, str]:
```

Comment (Rocketknight1) [reply trl/data_utils.py:63]: PR is here: https://github.com/huggingface/transformers/pull/35216

**Comment by August-murr** [comment]
@qgallouedec 
let me know if there is anything else I need to do.

File: trl/data_utils.py [hunk 5]
```diff
@@ -139,7 +149,9 @@ def apply_chat_template(example: dict[str, list[dict[str, str]]], tokenizer: Pre
 
 
 def maybe_apply_chat_template(
-    example: dict[str, list[dict[str, str]]], tokenizer: PreTrainedTokenizer
+    example: dict[str, list[dict[str, str]]],
+    tokenizer: PreTrainedTokenizer,
+    tools: Optional[list[Union[dict, Callable]]] = None,
```
Comment (qgallouedec) [inline trl/data_utils.py:154]: Coule you add a line for this new arg in the docstring as well?

**Review by qgallouedec — CHANGES_REQUESTED** [review]

**Comment by qgallouedec** [comment]
Looks good to me! Just one munir comment

**Comment by qgallouedec** [comment]
```python
from transformers import AutoProcessor
from trl import apply_chat_template

tokenizer = AutoProcessor.from_pretrained("trl-internal-testing/tiny-LlamaForCausalLM-3.2")

# Define dummy test tools
def get_current_temperature(location: str):
    """
    Gets the temperature at a given location.

    Args:
        location: The location to get the temperature for
    """
    return 22.0

# Define test case
test_case = {
    "prompt": [
        {"content": "Whats the temperature in London?", "role": "user"},
    ]
}
# Test with tools
result_with_tools = apply_chat_template(test_case, tokenizer, tools=[get_current_temperature])

print(result_with_tools["prompt"])
```


```
<|begin_of_text|><|start_header_id|>system<|end_header_id|>

Environment: ipython
Cutting Knowledge Date: December 2023
Today Date: 12 Dec 2024

<|eot_id|><|start_header_id|>user<|end_header_id|>

Given the following functions, please respond with a JSON for a function call with its proper arguments that best answers the given prompt.

Respond in the format {"name": function name, "parameters": dictionary of argument name and its value}.Do not use variables.

{
    "type": "function",
    "function": {
        "name": "get_current_temperature",
        "description": "Gets the temperature at a given location.",
        "parameters": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "The location to get the temperature for"
                }
            },
            "required": [
                "location"
            ]
        }
    }
}

Whats the temperature in London?<|eot_id|><|start_header_id|>assistant<|end_header_id|>


```


Nice!

**Review by qgallouedec — APPROVED** [review]
